levels.go

```diff
@@ -476,6 +481,58 @@ func (cd *compactDef) unlockLevels() {
 	cd.thisLevel.RUnlock()
 }
 
+type rangeWithSize struct {
+	start []byte
+	end   []byte
+	sz    int
+}
+
+func (cd *compactDef) getInputBounds() []rangeWithSize {
+	bounds := make([][]byte, 0, 2+len(cd.bot)+1)
+
+	// Key range of higher level may larger than lower level, so add them into bounds to simplify later computation.
+	kr := getKeyRange(cd.top)
```

Comment:
How about just use `[]byte{0}` as smallest, `[]byte{255}` as biggest?
Then we don't need to sort.

Reply:
`[]byte{255}` may not be the biggest key if user input contains `[]byte{255, 0}`... Use a dummy key range seems a little hack. 
If you really care about this `sort`, maybe we can compare `kr.left` and `cd.bot[0].Smallest()` then use the smaller one, compare `kr.right` and `cd.bot[len(cd.bot) - 1].Biggest()` then use the larger one.
IMO, this sort doesn't hurt the performance of compaction, because of the later compaction process will take a long time.

Reply:
The biggest doesn't matter actually, because we will not use it to split sub-compaction.
We can take only the smallest key of the second level to split range, ignore the biggest key.

Reply:
No. We use it in `compactBuildTables` as the terminate condition of the last compaction job. Otherwise, we need to add a special `if` here to use `nil` as end key of the last job...

Reply:
What about 
```go
func (cd *compactDef) getInputBounds() []rangeWithSize {
	bounds := make([][]byte, 0, len(cd.bot)+1)
	kr := getKeyRange(cd.top)

	smallest := y.KeyWithTs(y.ParseKey(cd.bot[0].Smallest()), math.MaxUint64)
	if y.CompareKeys(kr.left, smallest) < 0 {
		bounds = append(bounds, kr.left)
	} else {
		bounds = append(bounds, cd.bot[0].Smallest())
	}

	for i := 1; i < len(cd.bot)-1; i++ {
		smallest := y.KeyWithTs(y.ParseKey(cd.bot[i].Smallest()), math.MaxUint64)
		bounds = append(bounds, smallest)
	}

	biggest := y.KeyWithTs(y.ParseKey(cd.bot[len(cd.bot)-1].Biggest()), 0)
	if y.CompareKeys(kr.right, biggest) > 0 {
		bounds = append(bounds, kr.right)
	} else {
		bounds = append(bounds, biggest)
	}

	ranges := make([]rangeWithSize, 0, len(bounds))
	for i := 0; i < len(bounds)-1; i++ {
		start, end := bounds[i], bounds[i+1]
		sz := cd.sizeInRange(cd.top, cd.thisLevel.level, start, end)
		if len(cd.bot) != 0 {
			sz += cd.sizeInRange(cd.bot, cd.nextLevel.level, start, end)
		}
		ranges = append(ranges, rangeWithSize{start: start, end: end, sz: sz})
	}
	return ranges
}
```

Reply:
The nil check for `endKey` in `compactBuildTable` is already there.

Reply:
I mean here
https://github.com/coocood/badger/blob/947a1d48baa60ba627689257b443b73750681012/levels.go#L660-L668